### PR TITLE
Cache diagnostic ID strings in the AnalyzerConfigSet

### LINKED ISSUE #38426: .editorconfig is causing memory explosion in syntax trees
(FYI: This may be one of the root causes of #38330 reported by @davkean.)

I have a solution which is made up of multiple copies of the project-system repo (~400 projects in total). Looking a dump of the managed heap, nearly 10% of the managed heap (243mb of 2.3Gb) is consumed by duplicated strings held by `CSharpSyntaxTree.ParsedSyntaxTree`. The issue is that the project-system repo has an .editorconfig that turns diagnostics on/off and `Microsoft.CodeAnalysis.AnalyzerConfigSet.GetOptionsForSourcePath` does substring parsing of rules of the form "dotnet_diagnostic.<diagnostic>.severity". Since it does this _for each syntax tree in the solution_ (~6 million of them in this case), each substring parse is duplicated over and over and over again. It looks like if `AnalyzerConfig.Section` just did this substring parsing once and passed back the same string for everyone, this would eliminate the duplication.

### PR DESCRIPTION
**Pull request by agocke**
I'm not sure if this is exactly where we want to do the caching, so I've
taken the opportunity to make all the non-core AnalyzerConfig APIs internal.
There are no existing consumers depending on them, so this is not a breaking
change.

Fixes #38426

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
> There are no existing consumers depending on them, so this is not a breaking
change.

This API is going out in 16.3, so what about people taking a dependency on them then?

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs [hunk 1]
```diff
@@ -144,7 +156,8 @@ public AnalyzerConfigOptionsResult GetOptionsForSourcePath(string sourcePath)
                 TreeOptions.Builder treeBuilder,
                 AnalyzerOptions.Builder analyzerBuilder,
                 ArrayBuilder<Diagnostic> diagnosticBuilder,
-                string analyzerConfigPath)
+                string analyzerConfigPath,
+                SmallDictionary<ReadOnlyMemory<char>, string> diagIdCache)
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs]: `diagnosticsIdCache`

File: src/Compilers/Core/Portable/InternalUtilities/CharMemoryEqualityComparer.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        private CharMemoryEqualityComparer() { }
+
+        public bool Equals(ReadOnlyMemory<char> x, ReadOnlyMemory<char> y)
+        {
+            if (x.Length != y.Length)
+            {
+                return false;
+            }
+
+            var xSpan = x.Span;
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/InternalUtilities/CharMemoryEqualityComparer.cs]: This should all just be `x.Span.SequenceEquals(y.Span)`. That implementation of `SequenceEquals` is heavily optimized, particularly on CoreCLR.

Comment (agocke) [reply src/Compilers/Core/Portable/InternalUtilities/CharMemoryEqualityComparer.cs]: Nice, I didn't know that existed.

**Comment by agocke** [comment]
@jasonmalinowski Good question.

@333fred What's our plan for the APIs in 3.0? Can we take stuff back if we didn't mean to make it public. Some of this stuff wasn't supposed to be public surface area.

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs (hunk 1 above)
Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs]: Why not `diagnosticsIdentifierCache`?

Comment (jaredpar) [reply src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs]: Mostly the `diag` was bothering me. Fine with pretty much anything including the current.

**Comment by 333fred** [comment]
@agocke I mean, this will be a breaking change. We've tried not to make breaking changes were we can avoid them in the past.

**Comment by agocke** [comment]
@jaredpar This is the first update after we ship and it's really just an API dependency. I think the take back in such a short window is OK. It also wasn't available pre-16.3-p1, so the window here is really small.

**Comment by jaredpar** [comment]
I think the time frame is the key here. It's the first update hence our tolerance for breaking is much higher. There have been cases in the past where we shipped APIs and then immediately took them back. That minimizes the time they are available and hence reduces the likelyhood customers will be impacted.

If customers do report this change then it will be a positive sign that we should indeed make this public and take the long term maintenance hit.

**Comment by agocke** [comment]
Yes, I'm happy to respond to feedback here as well.

**Comment by 333fred** [comment]
Alright. Separately we need to update our PublicAPI files, I'm going to start a thread with infrastructure about doing this as part of our branching policy.

**Comment by agocke** [comment]
@333fred Any other comments?

**Pull request merged**